### ISSUE TITLE
Request throttled due to too many requests

### ISSUE THREAD

**Issue by smokedsalmonbagel** [toluaina/pgsync#92]
PGSync version: 
1.30
Postgres version: 
10.11
Elasticsearch version: 
7.7
Redis version: 

Python version: 

Problem Description: 
I get an error indicating that requests were throttled due to too many requests when doing the initial sync.  This seems like it is probably just that our ES instance is too small.  What is the best way to slow down the sync so that I can use my existing instance?

Error Message (if any): 

```
Exception AuthorizationException(403, '403 Request throttled due to too many requests

```
I am also getting quite a few errors intermittently like:
```
WARNING:elasticsearch: Undecodable raw error response from server: Extra data: line 1 column 5 (char 4)
```
I am assuming that this is due to some sort of unparseable error response coming back from ES due to the throttled requests. I am assuming these are related.

**Comment by toluaina** [comment]
Yes these 2 are possibly related. How large (roughly) is your dataset i.e rows or bytes?
You can try setting `ELASTICSEARCH_CHUNK_SIZE` to a large value although that could trigger other issues.
pgsync is ituned to sync data as fast as possible but you want the opposite.

You could also always clone the repo and introduce a small delay [here](https://github.com/toluaina/pgsync/blob/master/pgsync/sync.py#L678)

**Comment by smokedsalmonbagel** [comment]
The main table (index) is about 138k rows.  There are 4 other related tables in the document.  I set ELASTICSEARCH_CHUNK_SIZE to 1000 because I was having occasional issues at 2000.
Yes - essentially you made it _too_ fast :-) .  Good to know about the delay.  I will try that at least until we can scale up.

Tyler

**Comment by smokedsalmonbagel** [comment]
I am revisiting this.  It sounds like a good option at least for this error is to retry the _bulk call, probably with a back-off wait time.  I am trying to set something like that up.  Is there any reason this could be problematic?  Where is the best place to do that?  The exception is actually being raised [here](https://github.com/toluaina/pgsync/blob/8db39c21e8be01a2352b7dddfded9dc5b8bbb997/pgsync/sync.py#L744)

**Comment by toluaina** [comment]
Yes I could look into a backoff for the sync with a timeout interval.

But doesn't this suggest a configuration issue with the cluster itself?

**Comment by smokedsalmonbagel** [comment]
I am still working to confirm that based on our cluster logs.  
One issue I see is that a cluster may be able to handle the load of syncing records once the initial sync is complete.  But that  initial sync could be very considerable.

If the initial sync fails and pgsync restarts will it pick up where it left off or does pgsync need to start over?  I want to make sure I am understanding this process correctly.

**Comment by toluaina** [comment]
Good question. If the initial sync fails. It will restart all over. Breaking down thr initial sync into stages will require a bit of work

**Comment by smokedsalmonbagel** [comment]
Wanted to follow up with this.  I looked through our Elastic search metrics and it really didn't look like we were getting close to the writequeue max size but the cluster was having stability issues.  Usually they resulted in too many requests or 429 errors from our AWS ES instance.  We were running 3 nodes that were t2.small.elasticsearch size.   
We switched them to r5.large and have no issues.  Even though most ES docs say the JVMemory pressure can spike high it was causing an issue for us (I suspect OOM killer) - the node would just go away. 

Case and point - sizing up the ES nodes matter even if your database is small.
